plugins/storage/volume/linstor/src/main/java/org/apache/cloudstack/storage/datastore/driver/LinstorPrimaryDataStoreDriverImpl.java

```diff
@@ -0,0 +1,420 @@
+        final String rscName = "cs-" + vol.getUuid();
+        rscGrpSpawn.setResourceDefinitionName(rscName);
+        rscGrpSpawn.addVolumeSizesItem(vol.getSize() / 1024);
+
+        try
+        {
+            s_logger.debug("Linstor: Spawn resource " + rscName);
+            ApiCallRcList answers = linstorApi.resourceGroupSpawn(rscGrp, rscGrpSpawn);
+            if (ApiCallChecker.hasError(answers))
+            {
+                for (ApiCallRc answer : answers)
+                {
+                    s_logger.error(answer.getMessage());
+                }
+                throw new CloudRuntimeException("Linstor: Unable to spawn resource.");
+            }
+
+            List<ResourceWithVolumes> resources = linstorApi.viewResources(
+                Collections.emptyList(),
+                Collections.singletonList(rscName),
+                Collections.emptyList(),
+                null,
+                null,
+                null);
+            if (!resources.isEmpty() && !resources.get(0).getVolumes().isEmpty())
+            {
+                s_logger.info("Linstor: Created drbd device: " + resources.get(0).getVolumes().get(0).getDevicePath());
+                return resources.get(0).getVolumes().get(0).getDevicePath();
+            } else
+            {
+                s_logger.error("Linstor: viewResources didn't return resources or volumes.");
+                throw new CloudRuntimeException("Linstor: viewResources didn't return resources or volumes.");
+            }
+        } catch (ApiException apiEx)
+        {
+            s_logger.error("Linstor: ApiEx - " + apiEx.getMessage());
+            throw new CloudRuntimeException(apiEx);
+        }
+    }
+
+    @Override
+    public void createAsync(DataStore dataStore, DataObject vol, AsyncCompletionCallback<CreateCmdResult> callback)
+    {
+        String devPath = null;
+        String errMsg = null;
+        StoragePoolVO storagePool = _storagePoolDao.findById(dataStore.getId());
+
+        try
+        {
+            switch (vol.getType())
+            {
+                case VOLUME:
+                    VolumeInfo volumeInfo = (VolumeInfo) vol;
+                    VolumeVO volume = _volumeDao.findById(volumeInfo.getId());
+                    s_logger.debug("createAsync - creating volume");
+                    devPath = createResource((VolumeInfo) vol, storagePool);
+                    volume.setFolder("/dev/");
+                    volume.setPoolId(storagePool.getId());
+                    volume.setUuid(vol.getUuid());
+                    volume.setPath(vol.getUuid());
+
+                    _volumeDao.update(volume.getId(), volume);
+                    break;
+                case SNAPSHOT:
+                    s_logger.debug("createAsync - creating snapshot");
+                    break;
+                case TEMPLATE:
+                    s_logger.debug("createAsync - creating template");
+                    break;
```

Comment:
If the driver doesn't create snapshot / template here, better do not handle these cases.

Reply:
Hmm I wouldn't completely remove the cases, but I added an errMsg, so the callback gets an error answer,
If I would remove the cases, it would return an InvalidDataobject error, which it isn't